File: python/src/arkadia/ai/data/Node.py
from typing import Any, Dict, List, Optional
from .Schema import Schema, SchemaKind
from .Config import Config
from .Meta import Meta, MetaInfo
import json
import re


class Node(Meta):
    """
    Canonical runtime data object for AI.DATA.

    A Node always refers to a Schema which defines how data should be interpreted.

    Forms:
    - primitive:   value = 1,  "Aga",  True
    - record:      fields = { key: Node }
    - list:        elements = [Node, Node]

    Only values allowed by the Schema should be stored here.
    """

    def __init__(
        self,
        schema: Schema,
        *,
        name: Optional[str] = None,
        # primitive: 42, "A", True
        value: Any = None,
        # named: {"a": Node, "b": Node}
        fields: Optional[Dict[str, "Node"]] = None,
        # list: [Node,Node,...]
        elements: Optional[List["Node"]] = None,

        # list of comments associated with this node /* comment */ /* comment2 */
        comments: Optional[List[str]] = None,
        # Meta information (Runtime meta from data block) e.g. / @size=10 / -> {"size": 10}
        attr: Optional[Dict[str, Any]] = None,
        # Tags for meta #a #b #c
        tags: Optional[List[str]] = None,
    ):
        super().__init__(
            comments=comments, 
            attr=attr, 
            tags=tags)
        
        self.schema = schema
        self.name = name or ""

        self.value = value
        self.fields = fields or {}
        self.elements = elements or []


    # -----------------------------------------------------------
    # Introspection helpers
    # -----------------------------------------------------------

    @property
    def is_primitive(self):
        return self.schema is not None and self.schema.is_primitive

    @property
    def is_record(self):
        return self.schema is not None and self.schema.is_record

    @property
    def is_list(self):
        return self.schema is not None and self.schema.is_list

    # -----------------------------------------------------------
    #  Meta
    # -----------------------------------------------------------

    def clear_meta(self):
        self.clear_common_meta()

    def apply_meta(self, info: MetaInfo):
        """
        Applies ALL metadata, including constraints (!required).
        """
        # 1. Apply common stuff (meta dict, comments)
        self.apply_common_meta(info)

    # -----------------------------------------------------------
    # Dict
    # -----------------------------------------------------------

    def dict(self) -> Any:
        """
        Recursively converts the Node into a standard Python dictionary/list/primitive.
        Useful for debugging or interfacing with standard JSON libraries.
        """
        # print("Converting Node to dict...", self.schema, 
        #       "\n", 
        #       self.is_primitive,
        #       "\n", 
        #       self.is_list,
        #       "\n",
        #       "-" * 120)
        
        if self.is_primitive:
            return self.value

        if self.is_list:
            return [element.dict() for element in self.elements]

        if self.is_record:
            return {key: field_node.dict() for key, field_node in self.fields.items()}

        return self.value
    
    # -----------------------------------------------------------
    # json
    # -----------------------------------------------------------

    def json(self, indent: int = 2, colorize: bool = False) -> str:
        """
        Converts the Node to a JSON string.

        Args:
            indent (int): Number of spaces for indentation.
            colorize (bool): If True, applies ANSI colors to keys, strings, numbers, etc.
        """
        # 1. Convert to standard python dict
        data = self.dict()

        # 2. Dump to string
        json_str = json.dumps(data, indent=indent, ensure_ascii=False)

        if not colorize:
            return json_str

        # 3. Apply Colors (Simple Regex Tokenizer)
        # Using ANSI codes matching your Encoder class:
        # STRING = Green, NUMBER = Blue, BOOL = Magenta, NULL = Gray, KEY = Yellow

        RESET = "\033[0m"
        STRING = "\033[92m"
        NUMBER = "\033[94m"
        BOOL = "\033[95m"
        NULL = "\033[90m"
        KEY = "\033[93m"

        def replace_match(match):
            s = match.group(0)

            # Key ("key": )
            if re.match(r'^".*":$', s.strip()):
                return f"{KEY}{s}{RESET}"

            # String value ("value")
            if s.startswith('"'):
                return f"{STRING}{s}{RESET}"

            # Boolean
            if s in ("true", "false"):
                return f"{BOOL}{s}{RESET}"

            # Null
            if s == "null":
                return f"{NULL}{s}{RESET}"

            # Number (integers or floats)
            return f"{NUMBER}{s}{RESET}"

        # Regex to capture JSON tokens:
        # 1. Keys (quoted string followed by colon)
        # 2. String values
        # 3. Booleans/Null
        # 4. Numbers
        token_pattern = r'(".*?"\s*:)|(".*?")|\b(true|false|null)\b|(-?\d+(?:\.\d*)?(?:[eE][+\-]?\d+)?)'

        return re.sub(token_pattern, lambda m: replace_match(m), json_str)

    # -----------------------------------------------------------

    def encode(self, config: Config = {
        "indent": 2,
    }) -> str:
        """
        Debug fallback. Real encoder is in encoder.py.
        """
        from .Encoder import Encoder
        return Encoder(config).encode(self)

    def __repr__(self):
        """
        Technical debug representation.
        Format: <Node(KIND:type) value/len=... details...>
        """
        # 1. Type Info
        if self.schema:
            kind = self.schema.kind.name
            type_name = self.schema.type_name
            
            # Complex type display
            if self.is_list:
                el_type = self.schema.element.type_name if self.schema.element else "any"
                type_label = f"LIST[{el_type}]"
            elif self.is_record and type_name not in ("record", "any"):
                type_label = f"RECORD:{type_name}"
            else:
                type_label = f"{kind}:{type_name}"
        else:
            type_label = "UNKNOWN"


        header = f"<Node({type_label})"

        # 2. Content Info
        content = []
        
        if self.is_primitive:
            # Show value, truncated if necessary
            v = repr(self.value)
            if len(v) > 50: v = v[:47] + "..."
            content.append(f"val={v}")
            
        elif self.is_list:
            # Show count
            count = len(self.elements)
            content.append(f"len={count}")
            
        elif self.is_record:
            # Show keys summary
            keys = list(self.fields.keys())
            if len(keys) > 3:
                keys_str = ", ".join(keys[:3]) + ", ..."
            else:
                keys_str = ", ".join(keys)
            content.append(f"fields=[{keys_str}]")
        else:
            v = repr(self.value)
            if len(v) > 50: v = v[:47] + "..."
            content.append(f"val={v}")

        # 3. Meta Indicators (Concise)
        if self.comments:
            content.append(f"comments={len(self.comments)}")
        
        # Attributes
        current_attr = getattr(self, 'attr', getattr(self, 'meta', {}))
        if current_attr:
            content.append(f"attr={list(current_attr.keys())}")

        # Tags
        current_tags = getattr(self, 'tags', [])
        if current_tags:
            content.append(f"tags={current_tags}")

        # Assemble
        details_str = " " + " ".join(content) if content else ""
        return f"{header}{details_str}>"

File: typescript/src/models/Node.ts
import { Meta, MetaInfo, MetaProps } from './Meta';
import { Schema } from './Schema';

export interface NodeProps extends MetaProps {
    name?: string;
    value?: any;
    fields?: Record<string, Node>;
    elements?: Node[];
}

/**
 * Canonical runtime data object for AI.DATA.
 * A Node always refers to a Schema which defines how data should be interpreted.
 */
export class Node extends Meta {
    schema: Schema;
    name: string;
    value: any;
    fields: Record<string, Node>;
    elements: Node[];

    constructor(schema: Schema, props: NodeProps = {}) {
        super(props);
        this.schema = schema;
        this.name = props.name || "";
        this.value = props.value ?? null;
        this.fields = props.fields || {};
        this.elements = props.elements ? [...props.elements] : [];
    }

    // -----------------------------------------------------------
    // Introspection helpers
    // -----------------------------------------------------------

    get isPrimitive(): boolean { return this.schema && this.schema.isPrimitive; }
    get isRecord(): boolean { return this.schema && this.schema.isRecord; }
    get isList(): boolean { return this.schema && this.schema.isList; }

    // -----------------------------------------------------------
    // Meta
    // -----------------------------------------------------------

    clearMeta(): void {
        this.clearCommonMeta();
    }

    applyMeta(info: MetaInfo): void {
        // Applies ALL metadata (common stuff: meta dict, comments)
        this.applyCommonMeta(info);
    }



    // -----------------------------------------------------------
    // Conversion Methods (dict / json)
    // -----------------------------------------------------------

    /**
     * Recursively converts the Node into a standard JavaScript object/array/primitive.
     * Equivalent to Python's .dict() method.
     */
    dict(): any {
        if (this.isPrimitive) {
            return this.value;
        }

        if (this.isList) {
            return this.elements.map(element => element.dict());
        }

        if (this.isRecord) {
            const result: Record<string, any> = {};
            for (const [key, fieldNode] of Object.entries(this.fields)) {
                result[key] = fieldNode.dict();
            }
            return result;
        }

        return this.value;
    }

    /**
     * Converts the Node to a JSON string.
     * * @param indent Number of spaces for indentation.
     * @param colorize If true, applies ANSI colors to keys, strings, numbers, etc.
     */
    JSON(indent: number = 2, colorize: boolean = false): string {
        // 1. Convert to standard JS object
        const data = this.dict();

        // 2. Dump to string
        const jsonStr = JSON.stringify(data, null, indent);

        if (!colorize) {
            return jsonStr;
        }

        // 3. Apply Colors (Regex Tokenizer)
        // ANSI codes matching the Encoder class
        const C = {
            RESET: "\x1b[0m",
            STRING: "\x1b[92m", // Green
            NUMBER: "\x1b[94m", // Blue
            BOOL: "\x1b[95m",   // Magenta
            NULL: "\x1b[90m",   // Gray
            KEY: "\x1b[93m"     // Yellow
        };

        // Regex to capture JSON tokens:
        // Group 1: Keys ("key": )
        // Group 2: String values ("value")
        // Group 3: Booleans/Null
        // Group 4: Numbers
        const tokenPattern = /(".*?"\s*:)|(".*?")|\b(true|false|null)\b|(-?\d+(?:\.\d*)?(?:[eE][+\-]?\d+)?)/g;

        return jsonStr.replace(tokenPattern, (match) => {
            // Check based on content patterns
            
            // Key: Ends with ':' (ignoring whitespace) and starts with '"'
            if (/^".*":\s*$/.test(match)) {
                return `${C.KEY}${match}${C.RESET}`;
            }

            // String value: Starts with '"'
            if (match.startsWith('"')) {
                return `${C.STRING}${match}${C.RESET}`;
            }

            // Boolean
            if (match === "true" || match === "false") {
                return `${C.BOOL}${match}${C.RESET}`;
            }

            // Null
            if (match === "null") {
                return `${C.NULL}${match}${C.RESET}`;
            }

            // Number (Fallthrough)
            return `${C.NUMBER}${match}${C.RESET}`;
        });
    }
    
    // -----------------------------------------------------------
    // Debug / Representation
    // -----------------------------------------------------------

    /**
     * Technical debug representation.
     * Format: <Node(KIND:type) value/len=... details...>
     */
    toString(): string {
        // 1. Type Info
        let typeLabel = "UNKNOWN";
        if (this.schema) {
            const kind = this.schema.kind;
            const typeName = this.schema.typeName;

            if (this.isList) {
                const elType = this.schema.element ? (this.schema.element.typeName || "any") : "any";
                typeLabel = `LIST[${elType}]`;
            } else if (this.isRecord && typeName !== "record" && typeName !== "any") {
                typeLabel = `RECORD:${typeName}`;
            } else {
                typeLabel = `${kind}:${typeName}`;
            }
        }

        const header = `<Node(${typeLabel})`;
        const content: string[] = [];

        // 2. Content Info
        if (this.isPrimitive) {
            let v = String(this.value);
            if (typeof this.value === 'string') v = `"${v}"`;
            if (v.length > 50) v = v.substring(0, 47) + "...";
            content.push(`val=${v}`);
        } else if (this.isList) {
            content.push(`len=${this.elements.length}`);
        } else if (this.isRecord) {
            const keys = Object.keys(this.fields);
            let keysStr = "";
            if (keys.length > 3) {
                keysStr = keys.slice(0, 3).join(", ") + ", ...";
            } else {
                keysStr = keys.join(", ");
            }
            content.push(`fields=[${keysStr}]`);
        } else {
            // Fallback
            let v = String(this.value);
            if (v.length > 50) v = v.substring(0, 47) + "...";
            content.push(`val=${v}`);
        }

        // 3. Meta Indicators
        if (this.comments.length > 0) {
            content.push(`comments=${this.comments.length}`);
        }
        if (Object.keys(this.attr).length > 0) {
            content.push(`attr=[${Object.keys(this.attr).join(', ')}]`);
        }
        if (this.tags.length > 0) {
            content.push(`tags=[${this.tags.join(', ')}]`);
        }

        const detailsStr = content.length > 0 ? " " + content.join(" ") : "";
        return `${header}${detailsStr}>`;
    }


}